src/engagement_db_to_analysis/engagement_db_to_analysis.py

```diff
@@ -117,9 +117,54 @@ def _convert_messages_to_traced_data(user, messages_map):
 
     return messages_td
 
-#TODO: Fold messages by uid
-def _fold_messages_by_uid():
-    return None
+def _fold_messages_by_uid(user, message_traced_data):
+    """
+    Groups Messages TracedData objects into individual TracedData objects.
+
+    :param user: identifier of user running the pipeline.
+    :type user: str
+    :param message_traced_data: Messages TracedData objects to group.
+    :type message_traced_data: list of TracedData
+    :return: individual TracedData objects.
+    :rtype: list of individual TracedData objects.
+    """
+
+    individuals_td = []
+    individuals_uids = set()  # for uid lookup
+    for message_td in message_traced_data:
+
+        participant_uuid = message_td["participant_uuid"]
+        update_key = message_td["dataset"]
+
+        # Check if we already have the participant traced data in individuals
+        if message_td["participant_uuid"] in individuals_uids:
+
+            # if it exists
+            # update the existing individual traced data with the message_td
+            for ind_td in individuals_td:
```

Comment:
If individuals_td was a dict of uuid -> individual then you don't need to maintain a separate individuals_uids set or pay to loop over this list for every message

Reply:
Thanks wanted to retain individuals as a list of traced data but I think there we can have it as a dict too.
Refactored  to a dictionary of uid -> td objects i.e
` participants_traced_data = {'avf-participant-uuid': {s01e01: [msg1, msg2], age: [msg1]}`